storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1745 @@ buf_block_t *fsp_get_latched_xdes_page(
+    else return DB_CORRUPTION;
+
+    if (err) return err;
+    goto new_page;
+  }
+
+  /** Complete the page allocation from FREE extent descriptor
+  or XDES_FSEG/XDES_FREE_FRAG extent list */
+  void complete()
+  {
+    if (m_new_state == XDES_FREE)
+    {
+      if (m_need_segment)
+        return initialize_segment_complete();
+      return initialize_free_frag_complete();
+    }
+    if (m_new_state == XDES_FSEG)
+      return alloc_from_fseg_complete();
+    return alloc_from_free_frag_complete();
+  }
+
+  /** Assign the fragment slot of the index node.
+  This step should be done after removing the old page
+  because there is a possiblity that FRAGMENT ARRAY
+  could be full. */
+  void assign_frag_slot()
+  {
+    if ((!m_need_segment && m_new_state == XDES_FREE) ||
+        m_new_state == XDES_FULL_FRAG ||
+        m_new_state == XDES_FREE_FRAG)
+      fseg_set_nth_frag_page_no(m_inode, m_iblock,
+                                fseg_find_free_frag_page_slot(m_inode),
+                                m_new_page, m_mtr);
+  }
+
+  /** Restore the page modified during page allocation */
+  void restore_old_pages()
+  {
+    for (auto &it : m_old_pages)
+    {
+      buf_block_t *block=
+        m_mtr->get_already_latched(page_id_t{0, it.first},
+                                   MTR_MEMO_PAGE_SX_FIX);
+      ut_ad(block);
+      memcpy_aligned<UNIV_PAGE_SIZE_MIN>(
+        block->page.frame, it.second->page.frame, srv_page_size);
+    }
+  }
+};
+
+/** Steps for removing the page from old extent:
+
+(1) If the xdes_get_state(m_old_descr) is XDES_FSEG then
+   (1.1) If the number of used pages in descriptor is
+         FSP_EXTENT_SIZE then move the extent descriptor
+         from FSEG_FULL to FSEG_NOT_FULL
+   (1.2) If the number of used page in descriptor to 0 then
+         move the extent descriptor from FSEG_NOT_FULL to FSP_FREE
+   (1.3) Free the page and mark the XDES_FREE_BIT of the
+         respective page in current extent descriptor
+
+(2) If the xdes_get_stats(m_old_descr) is
+    XDES_FREE_FRAG / XDES_FULL_FRAG then
+
+    (2.1) If the number of used pages in extent descriptor
+          is FSP_EXTENT_SIZE then move the extent descriptor
+          from FSP_FULL_FRAG to FSP_FREE_FRAG list
+    (2.2) If the number of used pages in extent descriptor is 0 then
+          move the extent descriptor from FSP_FREE_FRAG to FSP_FREE
+    (2.3) Free the page and mark the XDES_FREE_BIT of the respective
+          page in current extent descriptor */
+class pageRemover final
+{
+private:
+  /* Header block for the tablespace */
+  buf_block_t *const m_header_block= nullptr;
+  /* Index node block */
+  buf_block_t *const m_iblock= nullptr;
+  /* Index node */
+  fseg_inode_t *const m_inode= nullptr;
+  /* offset of index node within index node page*/
+  uint16_t m_ioffset= 0;
+  /* Maximum free limit of the tablespace */
+  uint32_t m_free_limit= 0;
+  /* Extent size */
+  uint32_t m_extent_size= 0;
```

Comment:
This is largely duplicating `pageAllocator`. Do we really need both objects, or could both allocation and freeing be handled by the same class and object?